File: easy/6. Find Closest Value In BST/find_closest_value_in_bst.cpp
#include <iostream>

class BST {
public:
  int value;
  BST *left;
  BST *right;

  BST(int val);
  BST &insert(int val);
};

int findClosestValueInBst(BST *tree, int target) {
  int current_closest = tree->value;
  BST *current = tree;
  while (current != NULL) {
    // this means that the current value is closer to the target
    // than our current closest value
    if (std::abs(current->value - target) <
        std::abs(current_closest - target)) {
      current_closest = current->value;
    }
    // if the target is less than the current value, we go left
    if (current->value > target) {
      current = current->left;
    } else {
      current = current->right;
    }
  }
  return current_closest;
}

int main() { return 0; }

File: easy/6. Find Closest Value In BST/find_closest_value_in_bst.go
package main

import "math"

type BST struct {
	Value int

	Left  *BST
	Right *BST
}

func (tree *BST) FindClosestValue(target int) int {
	var current_closest int = tree.Value
	current_node := tree

  for current_node != nil {
    if math.Abs(float64(target - current_closest)) > math.Abs(float64(target - current_node.Value)) {
      current_closest = current_node.Value
    }
    if target < current_node.Value {
      current_node = current_node.Left
    } else {
      current_node = current_node.Right
    }
  }

	return current_closest
}
